File: python/examples/ex01_local_predict_loop.py
"""Example 01 -- local screen predict-loop: screenshot -> /predict -> execute -> repeat.

Purpose
    Drive YOUR OWN desktop with the stateless ``POST /v1/predict`` endpoint:
    grab a screenshot, ask Coasty for the next GUI action(s), execute them
    locally with pyautogui, and repeat while the model answers
    ``status == "continue"`` (capped by ``--max-steps``).

Flow
    1. Capture the primary monitor with mss and downscale it to 1280x720 with
       Pillow (both ship in the optional ``coasty[local]`` extra and are
       imported lazily, so CI never needs a display).
    2. POST /v1/predict with the downscaled screenshot AND the matching
       ``screen_width``/``screen_height`` -- see the coordinate-scaling
       pitfall explained at :func:`make_local_capture`.
    3. Execute the returned actions through ``coasty.executor.ActionExecutor``
       with :class:`coasty.PyAutoGuiBackend`, scaling coordinates back up to
       the real screen.
    4. Loop while ``status == "continue"``; stop on ``done``/``fail`` or when
       ``--max-steps`` is reached.

Endpoints used
    POST /v1/predict (scope ``predict``)

Estimated cost
    5 credits ($0.05) per /predict call at 1280x720 (exactly 1280x720 is NOT
    HD, so no +1 HD surcharge; no trajectory is attached here -- see ex03 for
    server-side trajectories). Worst case = ``--max-steps`` x 5 cr. The exact
    itemized estimate is COMPUTED via :mod:`coasty.cost` and printed before
    any spend; sandbox keys (``sk-coasty-test-*``) print "$0 (sandbox)".

Safety
    - pyautogui FAILSAFE stays ON (the default): fling the mouse into a
      screen corner to abort the loop instantly.
    - ``raw`` actions are never executed -- the shared executor logs and
      skips them.
    - Idempotency note: per the docs, ``Idempotency-Key`` is only supported
      on POST /runs, /workflows runs, /machines and snapshot -- POST /predict
      does not accept one. That is safe: failed predictions are charged then
      auto-refunded, so the shared client retries them freely. ex03 sends a
      unique per-step ``Idempotency-Key`` on session predicts, which DO
      support it in this cookbook's client.

Run it
    python examples/ex01_local_predict_loop.py "Open the calculator" --max-steps 5 --confirm
"""

from __future__ import annotations

import argparse
import base64
import importlib
import io
import sys
from collections.abc import Callable, Sequence
from dataclasses import dataclass
from typing import Any, Literal, cast

from coasty import (
    ActionBackend,
    ActionExecutor,
    CoastyClient,
    CoastyError,
    PyAutoGuiBackend,
    cost,
    env,
)
from coasty.types import CuaVersion, PredictStatus

# Send screenshots at 1280x720: the largest size that avoids the +1 cr HD
# surcharge (HD is STRICTLY w > 1280 OR h > 720) while staying readable.
TARGET_WIDTH = 1280
TARGET_HEIGHT = 720

FinalStatus = Literal["done", "fail", "max_steps"]


@dataclass(frozen=True)
class CaptureResult:
    """One captured (and possibly downscaled) screenshot plus its geometry.

    ``sent_*`` is the size of the image actually encoded for the API;
    ``real_*`` is the physical screen size. The scale factors map the
    API's returned coordinates (in SENT space) back onto the REAL screen.
    """

    screenshot_b64: str  # raw base64, no data: prefix
    sent_width: int
    sent_height: int
    real_width: int
    real_height: int

    @property
    def scale_x(self) -> float:
        return self.real_width / self.sent_width

    @property
    def scale_y(self) -> float:
        return self.real_height / self.sent_height


CaptureFn = Callable[[], CaptureResult]
"""Injected screenshot source -- tests swap in a fake; main() uses mss."""


@dataclass(frozen=True)
class StepRecord:
    """What one loop iteration predicted and executed."""

    step: int
    status: PredictStatus
    request_id: str
    executed: list[str]  # action_types handled by the executor, in order
    credits_charged: int


@dataclass(frozen=True)
class LoopResult:
    """Outcome of the whole predict loop."""

    status: FinalStatus
    steps: list[StepRecord]

    @property
    def total_credits(self) -> int:
        return sum(record.credits_charged for record in self.steps)


def make_local_capture(
    target_width: int = TARGET_WIDTH, target_height: int = TARGET_HEIGHT
) -> CaptureFn:
    """Build a real-screen capture function over mss + Pillow.

    Both libraries come from the optional ``coasty[local]`` extra and are
    imported lazily so importing this module never requires a display.

    COORDINATE-SCALING PITFALL: /predict (and /ground) return x/y in the
    pixel space of the screenshot you SENT. If you downscale the screenshot
    you MUST (a) pass the downscaled ``screen_width``/``screen_height`` in
    the request body, and (b) multiply the returned coordinates back up by
    ``real / sent`` per axis before clicking. Mixing spaces -- e.g. sending
    a 1280x720 image while claiming a 2560x1440 screen, or clicking the raw
    coordinates on a larger screen -- lands every click in the wrong place
    (typically the top-left quadrant). ``ActionExecutor(scale_x=, scale_y=)``
    does step (b) for you.
    """
    try:
        # Lazy: mss/Pillow are optional ([local] extra); cast keeps mypy
        # strict happy without needing their type stubs.
        mss_module = cast(Any, importlib.import_module("mss"))
        image_module = cast(Any, importlib.import_module("PIL.Image"))
    except ImportError as exc:  # pragma: no cover - depends on environment
        raise RuntimeError(
            "mss/Pillow are not installed. Install the local-automation extra: "
            'pip install "coasty[local]"'
        ) from exc

    def capture() -> CaptureResult:  # pragma: no cover - needs a real display
        with mss_module.mss() as screen:
            monitor = screen.monitors[1]  # 1 = primary monitor (0 = all merged)
            shot = screen.grab(monitor)
        image = image_module.frombytes("RGB", shot.size, shot.bgra, "raw", "BGRX")
        real_width, real_height = int(image.size[0]), int(image.size[1])
        sent_width, sent_height = real_width, real_height
        if real_width > target_width or real_height > target_height:
            # Downscale (never upscale) -- and remember: the request body
            # below must carry THESE dimensions, not the real ones.
            image = image.resize((target_width, target_height))
            sent_width, sent_height = target_width, target_height
        buffer = io.BytesIO()
        image.save(buffer, format="PNG")
        return CaptureResult(
            screenshot_b64=base64.b64encode(buffer.getvalue()).decode("ascii"),
            sent_width=sent_width,
            sent_height=sent_height,
            real_width=real_width,
            real_height=real_height,
        )

    return capture


def spend_gate(
    estimate: cost.CostEstimate,
    *,
    api_key: str,
    confirm: bool,
    title: str = "Estimated cost",
    emit: Callable[[str], None] = print,
) -> bool:
    """Print the itemized estimate; decide whether a billable call may run.

    Sandbox keys (``sk-coasty-test-*``) never bill -- they are labeled
    "$0 (sandbox)" and proceed. Live keys require ``--confirm`` or
    ``COASTY_CONFIRM_SPEND=1``. (Shared by ex02/ex03/ex05.)
    """
    sandbox = env.is_sandbox_key(api_key)
    emit(cost.format_estimate(estimate, title=title, sandbox=sandbox))
    if sandbox:
        emit("$0 (sandbox) -- sandbox keys never bill; proceeding.")
        return True
    if confirm or env.spend_confirmed():
        emit("Spend confirmed (--confirm / COASTY_CONFIRM_SPEND=1); proceeding.")
        return True
    emit("Refusing to spend on a live key: re-run with --confirm or set COASTY_CONFIRM_SPEND=1.")
    return False


def run_predict_loop(
    client: CoastyClient,
    capture: CaptureFn,
    backend: ActionBackend,
    instruction: str,
    *,
    max_steps: int = 10,
    cua_version: CuaVersion = "v3",
    emit: Callable[[str], None] = print,
) -> LoopResult:
    """The pure, testable core: capture -> predict -> execute, until terminal.

    ``capture`` and ``backend`` are injected so tests run headless (a fake
    capture + :class:`coasty.NullBackend`); main() wires up mss + pyautogui.
    """
    if max_steps < 1:
        raise ValueError("max_steps must be >= 1")
    steps: list[StepRecord] = []
    for step_number in range(1, max_steps + 1):
        shot = capture()
        result = client.predict(
            shot.screenshot_b64,
            instruction,
            cua_version=cua_version,
            # MUST match the screenshot actually sent (see make_local_capture).
            screen_width=shot.sent_width,
            screen_height=shot.sent_height,
        )
        prediction = result.data
        status: PredictStatus = prediction["status"]
        # Scale predicted coordinates (sent space) back to the real screen.
        executor = ActionExecutor(backend, scale_x=shot.scale_x, scale_y=shot.scale_y)
        executed = executor.execute_all(prediction["actions"])
        record = StepRecord(
            step=step_number,
            status=status,
            request_id=prediction["request_id"],
            executed=executed,
            credits_charged=prediction["usage"]["credits_charged"],
        )
        steps.append(record)
        emit(
            f"step {step_number}/{max_steps}: status={status} actions={executed} "
            f"({record.credits_charged} cr, request_id={record.request_id})"
        )
        if status != "continue":
            return LoopResult(status=status, steps=steps)
    emit(f"stopping: --max-steps {max_steps} reached before done/fail")
    return LoopResult(status="max_steps", steps=steps)


def build_estimate(max_steps: int, cua_version: CuaVersion) -> cost.CostEstimate:
    """Worst-case itemized estimate: ``max_steps`` predicts at 1280x720."""
    per_step = cost.estimate_predict(
        cua_version=cua_version, screen_width=TARGET_WIDTH, screen_height=TARGET_HEIGHT
    )
    return cost.CostEstimate(
        items=tuple(
            cost.CostItem(f"{item.label} x{max_steps} step(s)", item.credits * max_steps)
            for item in per_step.items
        )
    )


def main(argv: Sequence[str] | None = None) -> int:
    parser = argparse.ArgumentParser(
        description="Drive the local desktop via POST /v1/predict (screenshot loop)."
    )
    parser.add_argument("instruction", help="natural-language task, e.g. 'Open the calculator'")
    parser.add_argument("--max-steps", type=int, default=10, help="hard cap on loop iterations")
    parser.add_argument("--cua-version", choices=("v1", "v3", "v4"), default="v3")
    parser.add_argument(
        "--confirm", action="store_true", help="allow spending on a live (non-sandbox) key"
    )
    args = parser.parse_args(list(argv) if argv is not None else None)
    cua_version = cast(CuaVersion, args.cua_version)

    try:
        api_key = env.require_api_key()
    except env.MissingAPIKeyError as exc:
        print(f"error: {exc}", file=sys.stderr)
        return 2

    estimate = build_estimate(args.max_steps, cua_version)
    if not spend_gate(estimate, api_key=api_key, confirm=args.confirm, title="ex01 predict loop"):
        return 2

    try:
        capture = make_local_capture()
        # FAILSAFE stays True: moving the mouse into any screen corner raises
        # pyautogui.FailSafeException and aborts mid-action.
        backend = PyAutoGuiBackend(failsafe=True)
    except RuntimeError as exc:
        print(f"error: {exc}", file=sys.stderr)
        return 2

    try:
        with CoastyClient(api_key) as client:
            outcome = run_predict_loop(
                client,
                capture,
                backend,
                args.instruction,
                max_steps=args.max_steps,
                cua_version=cua_version,
            )
    except CoastyError as exc:
        print(f"error: {exc}", file=sys.stderr)  # str() includes the request_id
        if exc.request_id:
            print(f"request_id: {exc.request_id}", file=sys.stderr)
        return 1

    print(
        f"finished: {outcome.status} after {len(outcome.steps)} step(s); "
        f"{outcome.total_credits} credit(s) charged"
    )
    return 0 if outcome.status == "done" else 1


if __name__ == "__main__":
    sys.exit(main())

File: typescript/src/examples/ex01-local-predict-loop.ts
/**
 * Example 01 — Local screen predict-loop (browser target).
 *
 * Purpose: drive a live screen with the stateless `/predict` endpoint:
 * screenshot -> POST /v1/predict -> execute the returned actions -> repeat
 * while `status === "continue"`, capped by `--max-steps`.
 *
 * The "screen" here is a Playwright page pinned to a 1280x720 viewport:
 * `page.screenshot()` goes in, `page.mouse` / `page.keyboard` actions come
 * out. Playwright is OPTIONAL — the core loop only needs an injected
 * screenshot provider + an `ExecutorBackend` (src/coasty/executor.ts), and
 * the real Playwright wiring sits behind a dynamic import in `main()`.
 *
 * Endpoints: POST /v1/predict (5 credits per call at <=1280x720; the HD
 * surcharge does NOT apply because exactly 1280x720 is not HD).
 *
 * Estimated cost (computed at runtime via src/coasty/cost.ts and printed by
 * the spend gate): maxSteps x estimatePredictCredits({1280, 720}) =
 * 8 x 5 cr = 40 cr = $0.40 ceiling with the defaults — the loop usually
 * finishes sooner and only completed steps bill. Sandbox keys
 * (sk-coasty-test-*) never bill: the gate prints "$0 (sandbox)".
 *
 * Run it:
 *   npx tsx src/examples/ex01-local-predict-loop.ts \
 *     --url https://example.com --instruction "Click the More information link" \
 *     --max-steps 8 --confirm
 *
 * Env config: COASTY_API_KEY (required), COASTY_BASE_URL, EX01_URL,
 * EX01_INSTRUCTION, EX01_MAX_STEPS, COASTY_CONFIRM_SPEND=1 (instead of
 * --confirm).
 */
import { randomUUID } from 'node:crypto';
import path from 'node:path';
import { pathToFileURL } from 'node:url';
import { CoastyClient, type CreateCallOptions } from '../coasty/client.js';
import {
  estimatePredictCredits,
  formatEstimate,
  formatUsd,
  type EstimateLineItem,
} from '../coasty/cost.js';
import { getApiKey, isSandboxKey, spendConfirmed } from '../coasty/env.js';
import { CoastyError } from '../coasty/errors.js';
import { executeActions, type ExecutorBackend } from '../coasty/executor.js';
import { type CuaVersion } from '../coasty/types.js';

// ---------------------------------------------------------------------------
// COORDINATE SCALING PITFALL (read this before pointing the loop at anything)
// ---------------------------------------------------------------------------
// /predict returns coordinates in the pixel space of the screenshot you SENT,
// together with the screen_width/screen_height you declared. This example
// pins the Playwright viewport to EXACTLY 1280x720 and sends
// screen_width=1280 / screen_height=720, so model coordinates map 1:1 onto
// page.mouse coordinates (scaleX = scaleY = 1). Bonus: 1280x720 stays just
// under the HD surcharge (strictly >1280 wide OR >720 tall adds +1 credit
// per image; exactly 1280x720 is NOT HD).
//
// A DESKTOP backend (pyautogui / nut-js) differs in two ways:
//   1. You capture the PHYSICAL screen (e.g. 2560x1440, often DPI-scaled),
//      so you should downscale the PNG to <=1280x720 before sending, declare
//      the DOWNSCALED dimensions, and then multiply the returned x/y by
//      (real / sent) before moving the real mouse — that is exactly what
//      ExecuteOptions.scaleX / scaleY in src/coasty/executor.ts are for.
//   2. OS DPI scaling can make "screenshot pixels" differ from "mouse
//      coordinates" even without downscaling. Browser viewports avoid the
//      trap entirely: page.screenshot() captures CSS pixels that match
//      page.mouse coordinates 1:1 (at deviceScaleFactor 1).
// ---------------------------------------------------------------------------

export const SCREEN_WIDTH = 1280;
export const SCREEN_HEIGHT = 720;

export type PrintFn = (line: string) => void;

export const stdoutPrint: PrintFn = (line) => {
  process.stdout.write(`${line}\n`);
};

/** Returns one screenshot as raw base64 (NO `data:` prefix), as /predict expects. */
export type ScreenshotProvider = () => Promise<string>;

// ---------------------------------------------------------------------------
// PageLike — the minimal structural subset of a Playwright `Page` we need.
// Because the loop depends only on this interface (not on playwright types),
// the example compiles, lints and tests WITHOUT playwright installed.
// ---------------------------------------------------------------------------

export interface PageLikeMouse {
  click(
    x: number,
    y: number,
    options?: { button?: 'left' | 'right' | 'middle'; clickCount?: number },
  ): Promise<void>;
  move(x: number, y: number): Promise<void>;
  down(): Promise<void>;
  up(): Promise<void>;
  wheel(deltaX: number, deltaY: number): Promise<void>;
}

export interface PageLikeKeyboard {
  type(text: string): Promise<void>;
  press(key: string): Promise<void>;
}

export interface PageLike {
  screenshot(options?: { type?: 'png' }): Promise<Uint8Array>;
  mouse: PageLikeMouse;
  keyboard: PageLikeKeyboard;
}

/** Coasty key names (lowercase pyautogui style) -> Playwright key names. */
const PLAYWRIGHT_KEY_MAP: Readonly<Record<string, string>> = {
  enter: 'Enter',
  return: 'Enter',
  tab: 'Tab',
  esc: 'Escape',
  escape: 'Escape',
  space: 'Space',
  backspace: 'Backspace',
  delete: 'Delete',
  del: 'Delete',
  up: 'ArrowUp',
  down: 'ArrowDown',
  left: 'ArrowLeft',
  right: 'ArrowRight',
  home: 'Home',
  end: 'End',
  pageup: 'PageUp',
  pagedown: 'PageDown',
  ctrl: 'Control',
  control: 'Control',
  alt: 'Alt',
  shift: 'Shift',
  win: 'Meta',
  cmd: 'Meta',
  meta: 'Meta',
};

/** Map a Coasty/pyautogui key name ("ctrl", "enter", "a") to Playwright's. */
export function toPlaywrightKey(key: string): string {
  const mapped = PLAYWRIGHT_KEY_MAP[key.toLowerCase()];
  if (mapped !== undefined) return mapped;
  if (key.length === 1) return key; // letters, digits, punctuation
  return key.charAt(0).toUpperCase() + key.slice(1); // F1..F12, Insert, ...
}

/** Pixels scrolled per scroll "click" / amount unit. */
export const SCROLL_PIXELS_PER_UNIT = 100;

/** Wrap a PageLike into the ExecutorBackend the defensive executor drives. */
export function createPageBackend(page: PageLike): ExecutorBackend {
  return {
    async click(x, y, options): Promise<void> {
      await page.mouse.click(x, y, { button: options.button, clickCount: options.clicks });
    },
    async move(x, y): Promise<void> {
      await page.mouse.move(x, y);
    },
    async typeText(text): Promise<void> {
      await page.keyboard.type(text);
    },
    async keyPress(key): Promise<void> {
      await page.keyboard.press(toPlaywrightKey(key));
    },
    async keyCombo(keys): Promise<void> {
      // Playwright chords are "+"-joined: "Control+Shift+S".
      await page.keyboard.press(keys.map(toPlaywrightKey).join('+'));
    },
    async scroll(options): Promise<void> {
      if (options.x !== null && options.y !== null) await page.mouse.move(options.x, options.y);
      const pixels = options.amount * SCROLL_PIXELS_PER_UNIT;
      const deltaX =
        options.direction === 'left' ? -pixels : options.direction === 'right' ? pixels : 0;
      const deltaY =
        options.direction === 'up' ? -pixels : options.direction === 'down' ? pixels : 0;
      await page.mouse.wheel(deltaX, deltaY);
    },
    async drag(fromX, fromY, toX, toY): Promise<void> {
      await page.mouse.move(fromX, fromY);
      await page.mouse.down();
      await page.mouse.move(toX, toY);
      await page.mouse.up();
    },
    async wait(ms): Promise<void> {
      await new Promise((resolve) => setTimeout(resolve, ms));
    },
  };
}

/** Screenshot provider: PNG bytes from the page -> raw base64 (no data: prefix). */
export function createPageScreenshotProvider(page: PageLike): ScreenshotProvider {
  return async () => Buffer.from(await page.screenshot({ type: 'png' })).toString('base64');
}

// ---------------------------------------------------------------------------
// Core loop (pure: everything injected, fully testable offline)
// ---------------------------------------------------------------------------

export interface PredictLoopOptions {
  client: CoastyClient;
  instruction: string;
  screenshot: ScreenshotProvider;
  backend: ExecutorBackend;
  /** Hard cap on /predict calls (default 8). */
  maxSteps?: number;
  /** Declared screenshot dimensions (default 1280x720 — matches the viewport). */
  screenWidth?: number;
  screenHeight?: number;
  cuaVersion?: CuaVersion;
  /** real/sent coordinate factors for desktop backends (default 1 — see pitfall note). */
  scaleX?: number;
  scaleY?: number;
  /** Per-step Idempotency-Key factory (default: random, prefixed with the step). */
  idempotencyKey?: (step: number) => string;
  print?: PrintFn;
}

export type PredictLoopStatus = 'done' | 'fail' | 'max_steps';

export interface PredictLoopResult {
  status: PredictLoopStatus;
  /** Number of /predict calls made (each one bills). */
  stepsUsed: number;
  /** Sum of usage.credits_charged across all steps. */
  creditsCharged: number;
  /** The `fail` action's reason (or last reasoning) when status === "fail". */
  failReason: string | null;
  /** request_id of the last /predict response — quote it to support. */
  lastRequestId: string | null;
}

/**
 * The predict loop: screenshot -> /predict -> execute -> repeat while the
 * model says "continue". Stops on a `done`/`fail` response status, on a
 * terminal `done`/`fail` ACTION (defensive: some responses carry the marker
 * action while still saying "continue"), or at `maxSteps`.
 */
export async function runPredictLoop(options: PredictLoopOptions): Promise<PredictLoopResult> {
  const maxSteps = options.maxSteps ?? 8;
  const screenWidth = options.screenWidth ?? SCREEN_WIDTH;
  const screenHeight = options.screenHeight ?? SCREEN_HEIGHT;
  const print = options.print ?? stdoutPrint;
  const keyFor =
    options.idempotencyKey ?? ((step: number) => `ex01-step${String(step)}-${randomUUID()}`);

  let creditsCharged = 0;
  let lastRequestId: string | null = null;

  for (let step = 1; step <= maxSteps; step += 1) {
    const screenshot = await options.screenshot();
    // A fresh Idempotency-Key per step means a transport-level retry of THIS
    // step can never double-bill, while distinct steps stay distinct calls.
    const callOptions: CreateCallOptions = { idempotencyKey: keyFor(step) };
    const { data, meta } = await options.client.predict(
      {
        screenshot,
        instruction: options.instruction,
        cua_version: options.cuaVersion ?? 'v3',
        screen_width: screenWidth,
        screen_height: screenHeight,
      },
      callOptions,
    );

    lastRequestId = data.request_id !== '' ? data.request_id : meta.requestId;
    creditsCharged += data.usage.credits_charged;
    print(
      `[step ${String(step)}/${String(maxSteps)}] status=${data.status} ` +
        `actions=${String(data.actions.length)} credits=${String(data.usage.credits_charged)}` +
        (data.reasoning !== null ? ` reasoning=${data.reasoning}` : ''),
    );

    const results = await executeActions(data.actions, options.backend, {
      scaleX: options.scaleX ?? 1,
      scaleY: options.scaleY ?? 1,
      logger: print,
    });
    const terminal = results.find((result) => result.terminal !== null) ?? null;

    if (data.status === 'fail' || terminal?.terminal === 'fail') {
      const failReason = terminal?.detail ?? data.reasoning;
      print(`loop FAILED after ${String(step)} step(s): ${failReason ?? 'no reason given'}`);
      return { status: 'fail', stepsUsed: step, creditsCharged, failReason, lastRequestId };
    }
    if (data.status === 'done' || terminal?.terminal === 'done') {
      print(`loop DONE after ${String(step)} step(s).`);
      return { status: 'done', stepsUsed: step, creditsCharged, failReason: null, lastRequestId };
    }
  }

  print(`loop stopped: --max-steps=${String(maxSteps)} reached without done/fail.`);
  return {
    status: 'max_steps',
    stepsUsed: maxSteps,
    creditsCharged,
    failReason: null,
    lastRequestId,
  };
}

// ---------------------------------------------------------------------------
// Spend gate
// ---------------------------------------------------------------------------

export interface SpendGateOptions {
  apiKey: string;
  items: readonly EstimateLineItem[];
  /** True when the user passed --confirm. */
  confirmFlag: boolean;
  env?: NodeJS.ProcessEnv;
  print?: PrintFn;
}

/**
 * Print the itemized estimate, then decide whether billable calls may
 * proceed: sandbox keys always may ("$0 (sandbox)"); live keys require
 * `--confirm` or `COASTY_CONFIRM_SPEND=1`.
 */
export function ensureSpendApproved(options: SpendGateOptions): boolean {
  const print = options.print ?? stdoutPrint;
  const sandbox = isSandboxKey(options.apiKey);
  print(formatEstimate(options.items, { sandbox }));
  if (sandbox) {
    print('Spend gate: $0 (sandbox) — sandbox keys never bill; proceeding.');
    return true;
  }
  if (options.confirmFlag || spendConfirmed(options.env ?? process.env)) {
    print('Spend gate: confirmed — proceeding with billable calls.');
    return true;
  }
  print(
    'Spend gate: BLOCKED — this key bills real money. Re-run with --confirm ' +
      'or set COASTY_CONFIRM_SPEND=1 to proceed.',
  );
  return false;
}

// ---------------------------------------------------------------------------
// Optional Playwright wiring (dynamic import — never needed by tests/CI)
// ---------------------------------------------------------------------------

export interface PlaywrightPage extends PageLike {
  goto(url: string): Promise<unknown>;
}

interface PlaywrightBrowser {
  newPage(options?: { viewport: { width: number; height: number } }): Promise<PlaywrightPage>;
  close(): Promise<void>;
}

interface PlaywrightModule {
  chromium: { launch(options?: { headless?: boolean }): Promise<PlaywrightBrowser> };
}

export const PLAYWRIGHT_INSTALL_HINT =
  'Playwright is an OPTIONAL dependency of this example and is not installed.\n' +
  'Install it with:\n' +
  '  npm install -D playwright\n' +
  '  npx playwright install chromium\n' +
  'then re-run this example.';

/** Launch Chromium at a fixed viewport and navigate. Throws a clear hint when playwright is missing. */
export async function openPlaywrightPage(options: {
  url: string;
  headless: boolean;
  width?: number;
  height?: number;
}): Promise<{ page: PlaywrightPage; close: () => Promise<void> }> {
  // The module name is deliberately widened to `string` so tsc does not try
  // to resolve types for a package that may not be installed.
  const moduleName: string = 'playwright';
  let playwright: PlaywrightModule;
  try {
    playwright = (await import(moduleName)) as PlaywrightModule;
  } catch (cause) {
    throw new Error(PLAYWRIGHT_INSTALL_HINT, { cause });
  }
  const browser = await playwright.chromium.launch({ headless: options.headless });
  const page = await browser.newPage({
    viewport: { width: options.width ?? SCREEN_WIDTH, height: options.height ?? SCREEN_HEIGHT },
  });
  await page.goto(options.url);
  return { page, close: () => browser.close() };
}

// ---------------------------------------------------------------------------
// CLI
// ---------------------------------------------------------------------------

export interface Ex01Config {
  url: string;
  instruction: string;
  maxSteps: number;
  headless: boolean;
  confirm: boolean;
}

function intOption(name: string, raw: string | undefined, fallback: number): number {
  if (raw === undefined || raw.trim() === '') return fallback;
  const value = Number(raw);
  if (!Number.isInteger(value) || value < 1) {
    throw new Error(`${name} must be a positive integer, got: ${raw}`);
  }
  return value;
}

export function parseArgs(
  argv: readonly string[],
  env: NodeJS.ProcessEnv = process.env,
): Ex01Config {
  const config: Ex01Config = {
    url: env.EX01_URL ?? 'https://example.com',
    instruction: env.EX01_INSTRUCTION ?? 'Click the "More information..." link.',
    maxSteps: intOption('EX01_MAX_STEPS', env.EX01_MAX_STEPS, 8),
    headless: true,
    confirm: false,
  };
  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    const next = (): string => {
      i += 1;
      const value = argv[i];
      if (value === undefined) throw new Error(`missing value for ${String(arg)}`);
      return value;
    };
    switch (arg) {
      case '--url':
        config.url = next();
        break;
      case '--instruction':
      case '-i':
        config.instruction = next();
        break;
      case '--max-steps':
        config.maxSteps = intOption('--max-steps', next(), config.maxSteps);
        break;
      case '--headed':
        config.headless = false;
        break;
      case '--confirm':
        config.confirm = true;
        break;
      default:
        throw new Error(
          `unknown argument: ${String(arg)} (expected --url, --instruction, --max-steps, --headed, --confirm)`,
        );
    }
  }
  return config;
}

export function describeError(error: unknown): string {
  if (error instanceof CoastyError) {
    return `${error.code}: ${error.message} (request_id=${error.requestId ?? 'n/a'})`;
  }
  return error instanceof Error ? error.message : String(error);
}

export async function main(argv: readonly string[] = process.argv.slice(2)): Promise<number> {
  const config = parseArgs(argv);
  const apiKey = getApiKey();
  const perStep = estimatePredictCredits({
    screenWidth: SCREEN_WIDTH,
    screenHeight: SCREEN_HEIGHT,
  });
  const approved = ensureSpendApproved({
    apiKey,
    confirmFlag: config.confirm,
    items: [
      {
        label: `${String(config.maxSteps)} x POST /predict @ ${String(SCREEN_WIDTH)}x${String(SCREEN_HEIGHT)} (not HD)`,
        credits: config.maxSteps * perStep,
      },
    ],
  });
  if (!approved) return 1;

  const client = new CoastyClient({ apiKey });
  const { page, close } = await openPlaywrightPage({ url: config.url, headless: config.headless });
  try {
    const result = await runPredictLoop({
      client,
      instruction: config.instruction,
      screenshot: createPageScreenshotProvider(page),
      backend: createPageBackend(page),
      maxSteps: config.maxSteps,
    });
    stdoutPrint(
      `result: ${result.status} after ${String(result.stepsUsed)} step(s), ` +
        `${String(result.creditsCharged)} cr charged (${formatUsd(result.creditsCharged)}), ` +
        `last request_id=${result.lastRequestId ?? 'n/a'}`,
    );
    return result.status === 'fail' ? 1 : 0;
  } finally {
    await close();
  }
}

/** True when this module is the CLI entrypoint (vs being imported by tests). */
export function isCliEntry(moduleUrl: string): boolean {
  const entry = process.argv[1];
  if (entry === undefined) return false;
  return pathToFileURL(path.resolve(entry)).href.toLowerCase() === moduleUrl.toLowerCase();
}

if (isCliEntry(import.meta.url)) {
  void main()
    .then((code) => {
      process.exitCode = code;
    })
    .catch((error: unknown) => {
      console.error(describeError(error));
      process.exitCode = 1;
    });
}
